Helper/FeatureSwitch/Decider.php

```diff
@@ -29,20 +32,131 @@ class Decider extends AbstractHelper {
+    private $_fsRepo;
+
     /**
-     * @param Context $context
-     * @param Manager $manager
+     * @var FeatureSwitchFactory
+     */
+    private $_fsFactory;
+
+    /**
+     * @param Context              $context
+     * @param CoreSession          $coreSession
+     * @param State                $state
+     * @param Manager              $manager
+     * @param FeatureSwitchFactory $fsFactory
+     * @param FeatureSwitchRepository $fsRepo
      * @codeCoverageIgnore
      */
     public function __construct(
         Context $context,
-        Manager $manager
+        CoreSession $coreSession,
+        State $state,
+        Manager $manager,
+        FeatureSwitchRepository $fsRepo,
+        FeatureSwitchFactory $fsFactory
     ) {
         parent::__construct($context);
+        $this->_session = $coreSession;
+        $this->_state = $state;
         $this->_manager = $manager;
+        $this->_fsRepo = $fsRepo;
+        $this->_fsFactory = $fsFactory;
     }
 
+    /**
+     * This method returns if a feature switch is enabled for a user.
+     * The way this is computed is as follows:
+     * - Get feature switch id
+     * - Set if unset.
+     * - Add switch name as salt to ID and find md5 hash
+     * - Get first 6 digits of MD5 and divide by 0xffffff. Should be between 0 and 1.
+     * - Multiply previous value by 100
+     *   and compare with rolloutPercentage to decide if in bucket.
+     *
+     * @param string $switchName
+     * @param int $rolloutPercentage
+     * @return bool
+     */
+    private function _isInBucket(string $switchName, int $rolloutPercentage) {
+        $this->_session->start();
+        $boltFeatureSwitchId = $this->_session->getBoltFeatureSwitchId();
+        if (!$boltFeatureSwitchId) {
+            $boltFeatureSwitchId = uniqid("BFS", true);
+            $this->_session->setBoltFeatureSwitchId($boltFeatureSwitchId);
+        }
+        $saltedString = $boltFeatureSwitchId . "-" . $switchName;
+        $hash = md5($saltedString);
+        $hexStr = substr($hash, 0, 6);
+        $decEquivalent = hexdec($hexStr);
+
+        $hexMax = hexdec("0xffffff");
+        $position = $decEquivalent / (float) $hexMax * 100;
+        return $position < $rolloutPercentage;
+    }
+
+    private function _switchFromConst($switchDef) {
+        $switch = $this->_fsFactory->create();
+        $switch->setName($switchDef[Definitions::NAME_KEY]);
+        $switch->setValue($switchDef[Definitions::VAL_KEY]);
+        $switch->setDefaultValue($switchDef[Definitions::DEFAULT_VAL_KEY]);
+        $switch->setRolloutPercentage($switchDef[Definitions::ROLLOUT_KEY]);
+        return $switch;
+    }
+
+    /**
+     * This returns if the switch is enabled.
+     *
+     * @param string $switchName name of the switch
+     *
+     * @throws LocalizedException
+     * @return bool
+     */
+    public function isSwitchEnabled($switchName) {
+        $defaultDef = isset(Definitions::DEFAULT_SWITCH_VALUES[$switchName]) ?
```

Comment:
I missed this in the previous review, and it makes no different, but you *could* use the elvis operator here:
`$defaultDef = isset(Definitions::DEFAULT_SWITCH_VALUES[$switchName]) ?: null;`

Reply:
^ I missed the isset. Sorry, please ignore this.